src/rustdoc_gen.rs

```diff
@@ -455,13 +468,32 @@ impl RustdocGenerator for RustdocFromProjectRoot {
                 }
             }
         }).into_terminal_result()?;
+
+        Ok(Some(generate_data_request(
+            config,
+            CrateSource::ManifestPath { manifest },
+            crate_data,
+        )))
+    }
+
+    fn load_rustdoc(
+        &self,
+        config: &mut GlobalConfig,
+        generation_settings: super::data_generation::GenerationSettings,
+        cache_settings: super::data_generation::CacheSettings<()>,
+        data_request: &Option<CrateDataRequest<'_>>,
+    ) -> Result<VersionedStorage, TerminalError> {
+        let Some(data_request) = data_request else {
+            return Err(anyhow::anyhow!("`data_request` was unexpectedly `None`"))
+                .into_terminal_result();
+        };
+
```

Comment:
This method looks like it has the same shape for all the implementations of `RustdocGenerator` aside from the one that doesn't actually generate anything.

In all three remaining ones, it does essentially an `unwrap_or_err()` on `data_request` followed by a call to `generate_rustdoc()`.

This means we haven't really boiled this down to quite the right minimal abstraction yet.